### PR TITLE
Decouple logging from objects and base classes

### PR DESCRIPTION
**Pull request by happz**
Logging is baked into many classes, anything based on `tmt.utils.Common`
has methods like `debug` and `info` that handle verbosity checks,
indentations, colors, and logging into file *and* console. While having
those methods is absolutely handy, the fact code is embedded into every
class makes it much harder to parametrize it, influence, or redirect the
output. I believe this will become painfully visible when we start
cutting into parallelization where observability is crucial, and several
possible modes of output have been suggested already (everything in a
single stream/file vs per thread stream/file, "window" for each thread
on the same screen, interleaving, prefixing, screen or tmux integration,
multitail, etc.) - extending the current implementation of logging might
be near to impossible.

Hence this patch.

* Every object derived from `Common` is given a logger object when
  instantiated. This may very well be the very same logger every time,
  the dependency injection does not mean every `Common` instance would
  have its own logger instance.
* `Common` still provides helper methods in its namespace, proxying
  calls to the given logger.
* Python's own logging subsystem,
  https://docs.python.org/3/library/logging.html
  is very capable, and the patch uses it heavily, adding custom
  handlers, filters, log record details, and so on.
* When needed, code may spawn custom loggers to suit the needs of the
  situation. At this moment, the most needed functionality is
  "indentation", `shift` recognized by tmt's logging methods. In tmt's
  tree of runtime objects, some are expected to have their output
  shifted to the right, reflecting their position (`tmt` -> `Run` ->
  `Plan` -> `discover`/`provision`,...).

  In the future, there will be more features needed, e.g. to redirect
  particular output to different channels, or apply prefix to their
  lines, and so on.
* Distinction between command "output" and "logging" is now more
  visible, with output landing in stdout - no change there -, and all
  logging landing in stderr, and that's new.

This patch tries to implement all of the above, mimicking the current
observed behavior of tmt logging. There might be some minor changes, but
in general the output should be matching our current experience.

The most visible changes would probably be:

* many classes are now given `logger` argument, explicit
  `tmt.log.Logger` object to use for logging.
* logs at stderr - plenty of tests needed a change in what output they
  are checking (`2>&1 >/dev/null`), grepping stdout no longer works.

There are some caveats, though: there are still loggers out there that
tmt uses, but which are out of out control. The most prominent one is
the one from `fmf.utils`, fmf has it's own logging facilities, and may
break our nice little tidy world. The end goal is to teach fmf - and
others, when identified - to play nicely with others.

### CODE REVIEW

File: tests/core/debug/test.sh [hunk 1]
```diff
@@ -8,7 +8,7 @@ rlJournalStart
     rlPhaseEnd
 
     rlPhaseStartTest
-        rlRun "tmt run -ar prepare -ddd | tee output"
+        rlRun "tmt run -ar prepare -ddd 2>&1 >/dev/null | tee output"
```
[the remark this answers is not in the document]
Comment (happz) [reply tests/core/debug/test.sh:11]: Hmm, maybe I made a mistake, my goal was to swap stderr and stdout, but I see no third fd. I did not want to merge them into one stream, I wanted to redirect stderr to pipe.

In any case, feel free to propose a completely different approach, or several, and see also https://github.com/teemtee/tmt/issues/1565#issuecomment-1315923959 for more context/motivation. I’m not satisfied with swapping streams at all…

[the remark this answers is not in the document]
Comment (lukaszachy) [reply tests/core/debug/test.sh:11]: Yes, however we don't assert which tmt output is (expected) for stdout and which for stderr. Hence if we join them in each test we might introduce change by mistake in the future (and break user pipelines...)

[the remark this answers is not in the document]
Comment (happz) [reply tests/core/debug/test.sh:11]: IIUIC, `2>&1 >/dev/null` should actually redirect stderr to stdout, and redirect the original stdout to /dev/null, i.e. they shouldn't be merged into one stream, only stderr would go to `tee`.

[the remark this answers is not in the document]
Comment (thrix) [reply tests/core/debug/test.sh:11]: I agree with @happz, seems that is what is happening now here

[the remark this answers is not in the document]
Comment (thrix) [reply tests/core/debug/test.sh:11]: seems to be this is resolved @lukaszachy

File: tests/execute/result/basic.sh [hunk 2]
```diff
@@ -85,7 +85,7 @@ EOF
     rlPhaseEnd
 
     rlPhaseStartTest "Verbose execute prints result - abort case"
-        rlRun -s "tmt run --id \${run} --scratch --until execute tests tests -n /abort provision --how container execute -v" "2"
+        rlRun -s "tmt run --id \${run} --scratch --until execute tests tests -n /abort provision --how container execute -v 2>&1 >/dev/null" "2"
```
Comment (psss) [inline tests/execute/result/basic.sh:88]: Regarding the output catching: Here, for example we're using `rlRun -s` which gathers both `stdout` and `stderr`, so no adjustment is needed. For those cases where `command | tee output` is till used I'd suggest to migrate to `rlRun -s` instead.

For those cases when test script is piping the output to another command, for example `rlRun "tmt run | grep plan`, we could add `stderr` to `stdin` using `2>&1` in order to be able to inspect both outputs from the log. What do you think?

File: .pre-commit-hooks.yaml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 -   id: tmt-lint
     name: tmt lint
-    entry: bash -c "git ls-files --error-unmatch $(python3 -c 'import tmt; print(tmt.Tree(path=\".\").root)')/.fmf/version && tmt lint --source $@" PAD
+    entry: bash -c "git ls-files --error-unmatch $(python3 -c 'import tmt; import tmt.log; print(tmt.Tree(logger=tmt.log.Logger.create(), path=\".\").root)')/.fmf/version && tmt lint --source $@" PAD
```
[the remark this answers is not in the document]
Comment (thrix) [reply .pre-commit-hooks.yaml:3]: we discussed we will import `tmt.log` with `import tmt`

File: tests/execute/result/basic.sh (hunk 2 above)
Comment (thrix) [reply tests/execute/result/basic.sh:88]: We decided to move to `rlRun -s` where feasiable.

Rather, introduce a dedicated test for checking correct verbosity output, as makes sense.

**Comment by happz** [comment]
* [ ] review tests & `rlRun -s` & `tee`
* [ ] docs & contribute & logger from library
* [ ] `tmt` re-imports `Logger`

File: .pre-commit-hooks.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (happz) [reply .pre-commit-hooks.yaml]: Fixed in cdd986210f513c37ebabaddee3663c99632f4ca1

**Comment by happz** [comment]
@psss @lukaszachy @thrix hmmmm, after some attempts, I'm inclined to leave tests as they are now, get this patch merged, and then fix those pesky `-s` and `tee` issues in another patch. I discovered that while working on those stream swaps, I'm fixing `rlRun` calls I don't have to fix, just to use `-s` when all other `rlRun` calls around them use it, which presents two outcomes:

1. I won't do that, for the sake of patch clarity, which would leave tests in an inconsistent state, or
2. I would do it, and the patch would become even uglier, and I wouldn't change all tests, just those failing with logging changes, which would, again, leave tests in an inconsistent state.

In any case, I see a follow-up patch after the logging patch, finishing the job.

**Comment by psss** [comment]
> In any case, I see a follow-up patch after the logging patch, finishing the job.

You mean, finishing the code review first and only when we're good with the changes add another commit into this pull request which would cover the needed test changes?

**Comment by happz** [comment]
> > In any case, I see a follow-up patch after the logging patch, finishing the job.
> 
> You mean, finishing the code review first and only when we're good with the changes add another commit into this pull request which would cover the needed test changes?

Either that or an additional PR. In any case, this PR will be far from perfect: either it will change tests that are not affected by logging changes at all, i.e. changes unrelated to this PR, or it will change those that are & leave the rest to use `tee`, which leaves the "let's get rid of `tee` when we can use `-s`" action unfinished. So my proposal would be to leave tests as they are now, and add another PR targetting all tests that use `tee` and/or redirection, which would polish what's been introduced by this PR, and it'd have the "mandate" for touching all tests, not just those affected by logging changes.

**Review by thrix — CHANGES_REQUESTED** [review]

**Comment by thrix** [comment]
> > > In any case, I see a follow-up patch after the logging patch, finishing the job.
> > 
> > 
> > You mean, finishing the code review first and only when we're good with the changes add another commit into this pull request which would cover the needed test changes?
> 
> Either that or an additional PR. In any case, this PR will be far from perfect: either it will change tests that are not affected by logging changes at all, i.e. changes unrelated to this PR, or it will change those that are & leave the rest to use `tee`, which leaves the "let's get rid of `tee` when we can use `-s`" action unfinished. So my proposal would be to leave tests as they are now, and add another PR targetting all tests that use `tee` and/or redirection, which would polish what's been introduced by this PR, and it'd have the "mandate" for touching all tests, not just those affected by logging changes.

@happz I am fine with this approach myself.

@psss are you fine also?

File: tests/unit/__init__.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+                        expected_value, field_getter(
+                            record, field_name)))
+
+        assert False, f"""
+{message}:
+
+{chr(10).join(formatted_fields)}
+"""
+
+# Cannot find log record with these properties
```
[the remark this answers is not in the document]
Comment (happz) [reply tests/unit/__init__.py]: Fixed, and improved text of those messages a bit, in 9c4ed49fb8cb1921dcc55e7371ce5b6dffaebf46

File: tests/unit/conftest.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import _pytest.logging
+import pytest
+
+from tmt.log import Logger
+
+
+@pytest.fixture(name='root_logger')
+def fixture_root_logger(caplog: _pytest.logging.LogCaptureFixture) -> Logger:
+    logger = Logger.create(verbose=0, debug=0, quiet=False)
+    # caplog.set_level(logging.DEBUG, logger._logger.name)
```
[the remark this answers is not in the document]
Comment (happz) [reply tests/unit/conftest.py]: Fixed 9c4ed49fb8cb1921dcc55e7371ce5b6dffaebf46

**Comment by psss** [comment]
I'm ok with leaving the test changes for a separate commit or even a separate pull request. If we choose the second approach, I believe we should still resolve at least all tests failures so that other pull requests are not always failing. Or merge this one only once the second one is done.

**Comment by happz** [comment]
> I'm ok with leaving the test changes for a separate commit or even a separate pull request. If we choose the second approach, I believe we should still resolve at least all tests failures so that other pull requests are not always failing. Or merge this one only once the second one is done.

Yes, tests should be in good enough shape to be green. Just wanted to point out that a follow-up patch would be better to amend loose ends when it comes to the use of `tee` in tests, in general.

File: tests/unit/__init__.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+    def _report(message: str) -> None:
+        formatted_fields = [
+            f'    {field} == {value}'
+            for field, value in tests.items()
+            ]
+
+        for record in caplog.records:
+            for field_getter, field_name, op, expected_value in operators:
+                print(
```
[the remark this answers is not in the document]
Comment (thrix) [reply tests/unit/__init__.py]: thanks, looks a lot more readable to me

File: tests/execute/result/basic.sh (hunk 2 above)
Comment (thrix) [reply tests/execute/result/basic.sh:88]: I believe this is resolved.

We aim for green tests, improvments here for follow up

File: tmt/log.py [hunk 7]
```diff
@@ -0,0 +1,599 @@
...
+    @classmethod
+    def create(
+            cls,
+            raw_logger: Optional['Logger'] = None,
+            **verbosity_options: Any) -> 'Logger':
+        if raw_logger is None:
+            actual_logger = cls._normalize_logger(logging.getLogger('tmt'))
+
+        else:
+            actual_logger = raw_logger._logger
```
Comment (lukaszachy) [inline tmt/log.py:476]: How is this supposed to work when my app has its own logger and I want to wrap it? What do I miss?

```python
import logging
import tmt.log

my_logger = logging.getLogger('my_cool_app')
>>> tmt.log.Logger.create(my_logger)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".../site-packages/tmt/log.py", line 476, in create
    actual_logger = raw_logger._logger
AttributeError: 'Logger' object has no attribute '_logger'
```

FTR `tmt.log.Logger(my_logger)` works.

File: tmt/utils.py [hunk 8]
```diff
@@ -955,6 +902,9 @@ def _check_or_create_workdir_root_with_perms() -> None:
 
             # Create the workdir
             create_directory(workdir, 'workdir', quiet=True)
+
+        # print(f'{self.__class__.__name__}._workdir_init(id_={id_}) => {workdir}')
+        self._logger.add_logfile_handler(os.path.join(workdir, tmt.log.LOG_FILENAME))
```
Comment (lukaszachy) [inline tmt/utils.py:907]: Does this has to happen? When I provide my own logger I don't expect that common will add handlers there.

Comment (happz) [reply tmt/utils.py:913]: Added a comment in 7f0b5ba7ad99741430d476360a5f8ef74901bf96 - it probably does have to happen, and IMHO the fix should involve the refactoring of how workdir is created, this nasty logfile bit is just a consequence of lazy workdir creation.

File: tmt/log.py (hunk 7 above)
Comment (happz) [reply tmt/log.py]: Hopefully fixed in ab7486714b177845ff1494b5433cc29ca38f1982

File: tmt/steps/report/__init__.py [hunk 9]
```diff
@@ -4,6 +4,7 @@
 import click
 
 import tmt
+import tmt.result
```
Comment (thrix) [inline tmt/steps/report/__init__.py:7]: @happz this should be not needed, tmt exports `Result` in `__all__`

Comment (thrix) [reply tmt/steps/report/__init__.py:7]: line 85 can be changed to use directly `Result` I think

File: tmt/utils.py [hunk 10]
```diff
@@ -1277,7 +1234,9 @@ def shell_to_dict(variables: Union[str, List[str]]) -> EnvironmentType:
     return result
 
 
-def environment_to_dict(variables: Union[str, List[str]]) -> EnvironmentType:
+def environment_to_dict(logger: tmt.log.Logger,
+                        variables: Union[str,
```
Comment (thrix) [inline tmt/utils.py:1238]: we need to do something with this `autopep`, I cannot stop commenting on this, but here really autopep8 sucks :(

**Review by thrix — CHANGES_REQUESTED** [review]

File: tmt/steps/report/__init__.py (hunk 9 above)
Comment (happz) [reply tmt/steps/report/__init__.py:7]: This seems to be even unrelated change, I think I made it to make my IDE happy, but it's not touching any logging issue...

File: tmt/base.py [hunk 11]
```diff
@@ -2313,7 +2388,8 @@ def __init__(self,
                  *,
                  id_: Optional[str] = None,
                  tree: Optional[Tree] = None,
-                 context: Optional['tmt.cli.Context'] = None) -> None:
+                 context: Optional['tmt.cli.Context'] = None,
+                 logger: tmt.log.Logger) -> None:
```
Comment (psss) [inline tmt/base.py:2392]: Looking at this I realized: Having the `logger` as the last parameter somehow seems more natural to me... Not a big deal though, just wanted to share my feelings ;-)

File: tmt/utils.py [hunk 12]
```diff
@@ -426,9 +381,15 @@ class Common:
     # mostly fine.
     parent: Optional['Common'] = None
 
+    def inject_logger(self, logger: Optional[tmt.log.Logger] = None) -> None:
```
Comment (psss) [inline tmt/utils.py:384]: Please, add a short docstring.

Comment (psss) [reply tmt/utils.py:384]: Why is the `logger` parameter optional?

File: tmt/steps/prepare/__init__.py [hunk 13]
```diff
@@ -200,6 +205,7 @@ def go(self) -> None:
             # operations inside prepare plugins on the guest use the
             # prepare step config rather than provision step config.
             guest_copy = copy.copy(guest)
+            guest_copy._logger = guest._logger.clone().apply_verbosity_options(**self._options)
```
Comment (psss) [inline tmt/steps/prepare/__init__.py:208]: Shall we use `inject_logger` here?

File: tmt/cli.py [hunk 14]
```diff
@@ -163,23 +166,32 @@ def main(
...
-        common=tmt.utils.Common(),
-        fmf_context=tmt.utils.context_to_dict(context),
+        logger=logger,
+        common=tmt.utils.Common(logger=logger),
+        fmf_context=tmt.utils.context_to_dict(logger, context),
         steps=set(),
         tree=tree
         )
 
+    # print('tmt', click_contex.obj.logger)
```
Comment (psss) [inline tmt/cli.py:193]: Debug line?

File: tmt/base.py (hunk 11 above)
Comment (psss) [reply tmt/base.py:2392]: Hmmm... Looking more and more through the code, the feeling gets stronger and stronger...

File: tmt/base.py [hunk 15]
```diff
@@ -2228,14 +2303,14 @@ def stories(
 
         # Build the list, convert to objects, sort and filter
         stories = [
-            Story(node=story) for story
+            Story(node=story, logger=self._logger.descend()) for story
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/base.py:2306]: Good question. I would say no, hence `.descend()`, but it's up to us to decide. IIRC, this matches the current output, where each story is indented by this much, which `.descend()` achieves. As I see the semantics of `logger` not just here, but also in `tests()` and `plans()`, the logger is given to these methods for their logging, and to serve as a base for objects these methods return. By using the same for both, the method would log on the same level as each test, which doesn't seem right to me - or, rather, tests/plans/stories logging on the same level as `Tree` that *owns* them is what I think of as "not desirable".

File: tmt/base.py (hunk 11 above)
Comment (happz) [reply tmt/base.py:2392]: As long as we'd be dealing with keyword-only arguments, it's perfectly fine. I learned to state it as the first one, but no strong opinions here. If tmt's way would be "thou shalt declare logger as the last parameter", be it.

File: tmt/utils.py (hunk 12 above)
Comment (happz) [reply tmt/utils.py:384]: Relict of the distant past. Note the `assert` on the next line - I wanted to keep mypy happy, but I also wanted to catch code trying to sneak `None` as a logger. It should not be `None` anymore.

File: tmt/steps/prepare/__init__.py (hunk 13 above)
Comment (happz) [reply tmt/steps/prepare/__init__.py:208]: Correct, will fix.

File: tests/unit/test_utils.py [hunk 16]
```diff
@@ -755,13 +759,19 @@ def test_wait_deadline_already_passed():
     ticks = []
 
     with pytest.raises(WaitingTimedOutError):
-        wait(Common(), lambda: ticks.append(1), datetime.timedelta(seconds=-86400))
+        wait(
```
[the remark this answers is not in the document]
Comment (happz) [reply tests/unit/test_utils.py:762]: Fixed in 39891de6dbf9fb4e01471bda84147df2156d9eea

File: tmt/log.py [hunk 17]
```diff
@@ -0,0 +1,609 @@
+"""
+tmt's logging subsystem.
+
+Adds a layer on top of Python's own :py:mod:`logging` subsystem. This layer implements the desired
+verbosity and debug levels, colorization, formatting, verbosity inheritance and other features used
+by tmt commands and code.
+
+The main workhoses are :py:class:`Logger` instances. Each instance wraps a particular
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/log.py]: Fixed in 39891de6dbf9fb4e01471bda84147df2156d9eea

File: tmt/log.py [hunk 18]
```diff
@@ -0,0 +1,609 @@
...
+class _Formatter(logging.Formatter):
+    def __init__(self, fmt: str, apply_colors: bool = False) -> None:
+        super().__init__(fmt, datefmt='%H:%M:%S')
+
+        self.apply_colors = apply_colors
+
+        # TODO: this is an ugly hack, removing colors after they have been added...
+        # Wouldn't it be better to not add them at first place?
+        #
+        # This is needed to deal code that colorizes just part of the message, like
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/log.py]: Fixed in 39891de6dbf9fb4e01471bda84147df2156d9eea

File: tmt/log.py [hunk 19]
```diff
@@ -0,0 +1,609 @@
...
+        # and all that jazz. Convert it to actual message logging subsystem can emit.
+        if details is not None:
+            record.message = indent(
+                details['key'],
+                value=details['value'],
+                color=details.get('color', None) if self.apply_colors else None,
+                level=details.get('shift', 0)
+                )
+
+        # All other records are just the regular records,c reated by other pieces of code.
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/log.py]: Fixed in 39891de6dbf9fb4e01471bda84147df2156d9eea

File: tmt/log.py [hunk 20]
```diff
@@ -0,0 +1,609 @@
...
+    def descend(
+            self,
+            logger_name: Optional[str] = None,
+            extra_shift: int = 1
+            ) -> 'Logger':
+        """
+        Create a copy of this logger instance, but with a new raw logger.
+
+        New :py:class:`logging.Logger` instance is created from our raw logger, forming a
+        parent/child relationship betwen them, and it's then wrapped with ``Logger`` instance
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/log.py]: Fixed in 39891de6dbf9fb4e01471bda84147df2156d9eea

File: tmt/steps/report/__init__.py (hunk 9 above)
Comment (happz) [reply tmt/steps/report/__init__.py]: Fixed in 39891de6dbf9fb4e01471bda84147df2156d9eea

File: tmt/utils.py (hunk 10 above)
Comment (happz) [reply tmt/utils.py]: Fixed in 39891de6dbf9fb4e01471bda84147df2156d9eea

File: tmt/log.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (happz) [reply tmt/log.py]: Fixed in 39891de6dbf9fb4e01471bda84147df2156d9eea

File: tmt/log.py [hunk 21]
```diff
@@ -0,0 +1,609 @@
...
+        Parameters are passed directly to :py:class:`Logger`'s ``__init__()`` method.
+        """
+
+        return Logger(
+            actual_logger,
+            base_shift=base_shift,
+            verbosity_level=verbosity_level,
+            debug_level=debug_level,
+            quiet=quiet
+            )
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/log.py]: Fixed in 39891de6dbf9fb4e01471bda84147df2156d9eea

File: tmt/log.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (happz) [reply tmt/log.py]: Fixed in 39891de6dbf9fb4e01471bda84147df2156d9eea

File: tmt/utils.py (hunk 12 above)
Comment (happz) [reply tmt/utils.py]: Fixed in 39891de6dbf9fb4e01471bda84147df2156d9eea

File: tmt/steps/prepare/__init__.py (hunk 13 above)
Comment (happz) [reply tmt/steps/prepare/__init__.py]: Fixed in 39891de6dbf9fb4e01471bda84147df2156d9eea

File: tmt/cli.py (hunk 14 above)
Comment (happz) [reply tmt/cli.py]: Fixed in 39891de6dbf9fb4e01471bda84147df2156d9eea

File: tmt/base.py (hunk 11 above)
Comment (psss) [reply tmt/base.py:2392]: > If tmt's way would be "thou shalt declare logger as the last parameter", be it.

Thanks. Let's do it this way. The important parameters first and `logger` as the last one. When/if adding new parameters in the future `logger: tmt.log.Logger) -> None` will stay nicely untouched as the last line together with the return value.

Comment (happz) [reply tmt/base.py:2392]: One thing to resolve before I dive into moving parameters around: consider methods and functions with optional parameters:

```python
def environment_files_to_dict(
        env_files: Iterable[str],
        root: Optional[str] = None) -> EnvironmentType:
```

With logger as the always-first parameter, it looks like this, nice & tidy:

```python
def environment_files_to_dict(
        logger: tmt.log.Logger,
        env_files: Iterable[str],
        root: Optional[str] = None) -> EnvironmentType:
```

In the current state, `logger` cannot be truly last, Python would not allow that because `root` is a keyword argument with a default value, so we have two options:

```python
# last-but-not-so-last - this I find especially confusing
def environment_files_to_dict(
        env_files: Iterable[str],
        logger: tmt.log.Logger,
        root: Optional[str] = None) -> EnvironmentType:

# Turn the function into the keyword-only mode, then the logger can be last while other parameters keep their default values
def environment_files_to_dict(
        *,
        env_files: Iterable[str],
        root: Optional[str] = None,
        logger: tmt.log.Logger,) -> EnvironmentType:
```

Comment (psss) [reply tmt/base.py:2392]: I think turning functions with optional arguments to keyword-only mode makes sense and does not hurt much. It's a fair price for having the essential parameters at the beginning.

Comment (happz) [reply tmt/base.py:2392]: Ok, I'll add a commit tomorrow moving logger to the last position.

(`logger` *is* essential too O:)

Comment (psss) [reply tmt/base.py:2392]: > (`logger` _is_ essential too O:)

Yeah, sure, agreed, no doubt about it :)

File: tmt/log.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (psss) [reply tmt/log.py]: Good, thanks!

File: tmt/base.py (hunk 11 above)
Comment (happz) [reply tmt/base.py:2411]: First attempt, e3fa4f9f1e26099c13e35e6706b5599296c35f11

File: tmt/base.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (psss) [reply tmt/base.py:2320]: Sorry, probably a misunderstanding? I meant using `logger` instead of `self._logger` for descending:

```suggestion
            Story(node=story, logger=logger.descend()) for story
```

To be consistent with the way it's done for `tests()` and `plans()` above. Basically do not ignore `stories(..., logger)`.

File: tmt/utils.py (hunk 8 above)
Comment (psss) [reply tmt/utils.py:913]: Thanks for providing a detailed explanation. I guess that's all we can do here now.

File: tmt/base.py (hunk 11 above)
Comment (psss) [reply tmt/base.py:2411]: Thanks much! Changes look good.

**Review by thrix — APPROVED** [review]
LGTM for me. Awesome work.

File: tmt/log.py (hunk 7 above)
Comment (psss) [reply tmt/log.py]: Seems to be working fine now.

File: tmt/base.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (happz) [reply tmt/base.py]: Oh, right. That's even wrong, `self._logger` shouldn't be used, `logger` is the only right answer there. Fixed in 38c3be9b8b0789fae775146894cc614438101bb0

[the remark this answers is not in the document]
Comment (psss) [reply tmt/base.py]: Good.

File: tests/unit/__init__.py [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+
+       assert_log(message='everything went well', levelno=logging.INFO)
+       assert_log(message='things broke down', levelno=logging.ERROR)
+       assert_log(message=MATCH('user .+ logged in'), levelno=logging.INFO)
+
+    :param caplog: Pytest's `caplog` fixture.
+    :param evaluator: a callable reducing a given list of booleans into a single boolean. It is
+        used to evaluate whether the search for matching record was successfull: each record is
+        tested, and results of these per-record tests are passed to `evaluator` for the final
+        decision.
```
Comment (psss) [inline tests/unit/__init__.py:73]: Please, add a description of the `not_` parameter. Or perhaps we could rename it to `present` or `should_be_present`, `expected` or something like this?

File: tests/unit/test_logging.py [hunk 23]
```diff
@@ -0,0 +1,184 @@
...
+        # (
+        #   logger verbosity - corresponds to -v, -vv, -vvv CLI options,
+        #   message verbosity - `level` parameter of `verbosity(...)` call,
+        #   expected outcome of `VerbosityLevelFilter.filter()` - returns integer!
+        # )
+        (0, 0, 1),
+        (1, 0, 1),
+        (2, 0, 1),
+        (3, 0, 1),
+        (4, 0, 1),
```
Comment (psss) [inline tests/unit/test_logging.py:75]: There are no `verbose(level=0)` messages, so perhaps these five can be removed?

Comment (happz) [reply tests/unit/test_logging.py:75]: I included those for the sake of completeness, to venture into the territory of corner cases.

File: tests/unit/__init__.py (hunk 22 above)
Comment (happz) [reply tests/unit/__init__.py:73]: Changed in 6d14224c36569188909070dbdd7f1a5fe7fcbf6f

File: tests/unit/test_logging.py (hunk 23 above)
Comment (happz) [reply tests/unit/test_logging.py]: Dropped in 6d14224c36569188909070dbdd7f1a5fe7fcbf6f

**Review by psss — APPROVED** [review]
Thanks a lot for preparing this, @happz!

**Review by lukaszachy — APPROVED** [review]

**Comment by psss** [comment]
The beast has been merged! \o/ Thanks much to all, especially @happz for the patience!

**Comment by t184256** [comment]
Adding a mandatory keyword argument breaks API both ways.

How am I supposed to construct `tmt.Tree` in a way compatible with tmt both before and after this change?

**Comment by t184256** [comment]
I'm afraid that conflicting with 1.21, waiting until 1.21 hits stable and then bumping a dependency to 1.21 isn't a great strategy.

Second to not breaking API in the first place, what's the next best thing that can be done? Should I conditionalize on `tmt.log` existence? Can affected classes initialize a logger of its own when a logger isn't passed to them?

**Comment by happz** [comment]
> > If they are spawned under the hood of your work with `tmt.Tree`, they should be fine. If you create instances of classes like `tmt.base.Test` on your own, you need to supply a logger instance.
> 
> WDYM "they should be fine"?

If `tmt.base.Test` instance is created by `tmt.Tree`, it should get a logger instance derived from the one given to `tmt.Tree`. Your example is of the latter case, i.e. creating an instance of a base class on your own, therefore you need to provide a logger instance.

```python
(tmt) [pts-3:0]: happz@nest [main] ~/git/tmt $ python
Python 3.10.8 (main, Nov 14 2022, 00:00:00) [GCC 11.3.1 20220421 (Red Hat 11.3.1-3)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import logging
>>> logger = logging.getLogger()
>>> import tmt.log
>>> import tmt.base
>>> tmt.base.Tree(logger=tmt.log.Logger.create(logger)).plans()
[<tmt.base.Plan object at 0x7f922f4e2890>, <tmt.base.Plan object at 0x7f922dbf8e50>, <tmt.base.Plan object at 0x7f922dbf91b0>, <tmt.base.Plan object at 0x7f922dbf9840>, <tmt.base.Plan object at 0x7f922dbf9d80>, <tmt.base.Plan object at 0x7f922dbfa440>, <tmt.base.Plan object at 0x7f922dbfa8f0>, <tmt.base.Plan object at 0x7f922dbfaef0>, <tmt.base.Plan object at 0x7f922dbfb490>, <tmt.base.Plan object at 0x7f922dbfb970>, <tmt.base.Plan object at 0x7f922dbfbfd0>, <tmt.base.Plan object at 0x7f922da54550>, <tmt.base.Plan object at 0x7f922da54b20>]
>>> 
```

**Comment by psss** [comment]
> Can affected classes initialize a logger of its own when a logger isn't passed to them?

That was my original idea as well. Not for all classes though. I'm still pondering whether we should not make just one exception, at least for this super-common `tmt.Tree()` use case, and create a default logger there if not provided by user.

**Comment by happz** [comment]
> > Can affected classes initialize a logger of its own when a logger isn't passed to them?
> 
> That was my original idea as well. Not for all classes though. I'm still pondering whether we should not make just one exception, at least for this super-common `tmt.Tree()` use case, and create a default logger there if not provided by user.

The wrapper I promised to work on, to modify the signature & auto-create a logger when running in interactive mode, still stands. I do believe it's a better option than making logger optional in `Tree`, which would fool linters & open the door for human mistakes.
